src/ahp/prover.rs

```diff
@@ -457,29 +440,79 @@ impl<F: PrimeField> AHPForR1CS<F> {
                 *a *= b;
                 *a -= c * d;
             });
-        let rhs = r_alpha_evals.interpolate();
-        let q_1 = mask_poly.polynomial() + &rhs;
-        end_timer!(q_1_time);
+        let outer_poly = r_alpha_evals.interpolate_by_ref();
+        end_timer!(outer_poly_time);
+
+        let z_1_time = start_timer!(|| "Compute z_1 poly");
+
+        // In order to re-use the outer_poly_evals for computing z_1 (because z_1 shall be
+        // computed over domain H, while outer_poly_evals are on mul_domain), we ensure that
+        // H is a sub-domain of mul_domain.
+        assert_eq!(mul_domain.size() % domain_h.size(), 0);
+        let step = mul_domain.size()/domain_h.size();
+        assert_eq!(domain_h.group_gen(), mul_domain.group_gen().pow(&[step as u64]));
+
+        let outer_poly_evals_on_H = EvaluationsOnDomain::from_vec_and_domain(
+            r_alpha_evals.evals.clone().into_iter().step_by(step).collect(),
+            domain_h.clone()
+        );
+        let mut z_1 = BoundaryPolynomial::from_coboundary_polynomial_evals(
+            outer_poly_evals_on_H
+        ).unwrap().polynomial();
+
+        // Let b = zk_bound. This blinding polynomial is of type
+        // (c_0 + ... c_(2b-1) * X ^ (2b - 1) )
```

Comment:
Maybe worth to be more verbose on the reason:
the boundary poly is queried twice as much as the others.